pkg/util/condition/condition.go

```diff
@@ -0,0 +1,172 @@
+package condition
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	shipper "github.com/bookingcom/shipper/pkg/apis/shipper/v1alpha1"
+	diffutil "github.com/bookingcom/shipper/pkg/util/diff"
+)
+
+type Condition interface{}
+
+type ConditionDiff struct {
+	c1, c2 Condition
+}
+
+var _ diffutil.Diff = (*ConditionDiff)(nil)
+
+func NewConditionDiff(c1, c2 Condition) *ConditionDiff {
+	return &ConditionDiff{
+		c1: c1,
+		c2: c2,
+	}
+}
+
+func (cd *ConditionDiff) IsEmpty() bool {
+	if cd == nil || (condIsNil(cd.c1) && condIsNil(cd.c2)) {
+		return true
+	}
+	if condIsNil(cd.c1) || condIsNil(cd.c2) {
+		return false
+	}
+	return condEqual(cd.c1, cd.c2)
+}
+
+func (cd *ConditionDiff) String() string {
+	if cd.IsEmpty() {
+		return ""
+	}
+	var c1str, c2str string
+	if cd.c1 != nil {
+		c1str = condStr(cd.c1)
+	}
+	if cd.c2 != nil {
+		c2str = condStr(cd.c2)
+	}
+	return fmt.Sprintf("[%s] -> [%s]", c1str, c2str)
+}
+
+func condEqual(c1, c2 Condition) bool {
+	if condIsNil(c1) && condIsNil(c2) {
+		return true
+	}
+	if condIsNil(c1) || condIsNil(c2) {
+		return false
+	}
+	// Clumsy :-(
+	switch c1t := c1.(type) {
```

Comment:
I don't mind this for right now. If this lives on, though, the failure mode is gonna be someone introducing a new type whose values will always be different. I think we should fail as loud as possible (a panic, perhaps?) when that happens, to save someone's debugging time.